### PR TITLE
💩  Deprecated: Add edge case dataset and tests to `test-classify-diabetes.R`

### LINKED ISSUE #281: Start filling in explicit tests in `test-classify-diabetes.R`
See for example `test-prepare-lpr.R` file.

So that we know for certain which IDs should be correctly classified given a specific input data in the registers.

### LINKED ISSUE #303: Create fake case data set that contains potentially real cases/edge cases
name like `fake_cases`? 🤷

### PR DESCRIPTION
**Pull request by Aastedet**
## Description

- Added a dataset of 23 edge cases and expected outputs from these to `test-classify-diabetes.R`
- Added a `edge_case_tests()` function  to `test-classify-diabetes.R` with error messages depending on which case failed and how. I've verified the testing function as much as possible, but this will be easier to do when the `classify_diabetes()` function is updated and produces at least some correct outputs.
- Added a vignette describing each edge case in detail.
- Removed the old, commented-out tests in `test-classify-diabetes.R`

Closes #303, closes #281 

## Checklist

- [ ] Ran `just run-all`
- [ ] If docs were added, Markdown is formatted

### CODE REVIEW

File: R/create_edge_case_dataset.R [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+# Inclusion criteria dictionary (cases 12 - 16):
+
+
+
+# Censoring criteria dictionary (cases 17 - 21):
+
+# 1. bef: Demographics table -------------------------------------------------------------------------
+
+bef_tbl <- tibble::tribble(
+  ~pnr,                               ~koen, ~foed_dato,
```
[the remark this answers is not in the document]
Comment (lwjohnst86) [reply R/create_edge_case_dataset.R]: Yea, formatting is turned off for `tribble()` code (since it is non-standard structure).

File: R/test-cases-data.R [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+  osdc_test_data <- list(
+    bef = bef_tbl,
+    lmdb = lmdb_tbl,
+    lpr_adm = lpr_adm_tbl,
+    lpr_diag = lpr_diag_tbl,
+    kontakter = kontakter_tbl,
+    diagnoser = diagnoser_tbl,
+    sysi = sysi_tbl,
+    sssy = sssy_tbl,
+    lab_forsker = lab_forsker_tbl
```
Comment (lwjohnst86) [inline R/test-cases-data.R:295]: Add a final dataset that contains the expected output.

File: R/test-cases-data.R [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+  "14_t2d_gldF_diagF_hba1cT_podF",       "20130401",    "NPU27300",    56,
+  "21_nodm_female_pregnancyT",           "19980801",    "NPU27300",    55,
+  "21_nodm_female_pregnancyT",           "19990201",    "NPU27300",    55,
+  "21_nodm_female_pregnancyT",           "20230801",    "NPU27300",    55,
+  "21_nodm_female_pregnancyT",           "20240201",    "NPU27300",    55
+)
+
+  # Combine all tibbles into a named list -------------------------------------------------------------------------
+
+  osdc_test_data <- list(
```
Comment (lwjohnst86) [inline R/test-cases-data.R:286]:
```suggestion
  list(
```

File: R/test-cases-data.R [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+    lpr_adm = lpr_adm_tbl,
+    lpr_diag = lpr_diag_tbl,
+    kontakter = kontakter_tbl,
+    diagnoser = diagnoser_tbl,
+    sysi = sysi_tbl,
+    sssy = sssy_tbl,
+    lab_forsker = lab_forsker_tbl
+  )
+
+  return(osdc_test_data)
```
Comment (lwjohnst86) [inline R/test-cases-data.R:298]:
```suggestion
```

**Comment by signekb** [comment]
Add this test data to the `test-classify-diabetes.R` as well 👍

**Review by lwjohnst86 — CHANGES_REQUESTED** [review]

File: tests/testthat/test-classify-diabetes.R [hunk 5]
```diff
@@ -1,254 +1,410 @@
...
+  # Specify expected outputs (included cases):
+  expected_included <- create_expected_inclusions()
+
+  # First: ensure that there is exactly one row per pnr in the actual output:
+  test_that("Output should contain exactly one row per pnr", {
+    expect_true(length(unique(actual_included$pnr)) == nrow(actual_included))
+  })
+
+  # Iterate over the full list of pnr values from the input data
+  for (current_pnr in edge_case_data$bef$pnr) {
```
Comment (lwjohnst86) [inline tests/testthat/test-classify-diabetes.R:374]: I'm not entirely sure what you are trying to do here, and my guess (could be wrong) is that this is ChatGPT "assisted" code? ChatGPT does love loops...

This can be even simpler: `expect_identical(classify_diabetes(...), expected_cases)`

**Review by lwjohnst86 — CHANGES_REQUESTED** [review]
Some comments to start, since it will result in some major changes.

There are some formatting things in the vignette that I think I will manually fix myself, if that's ok

File: vignettes/edge-case-description.qmd
Comment (lwjohnst86) [inline vignettes/edge-case-description.qmd:1]: Most of should probably be treated as data and the doc generated from it, since it is pretty structured. I can do that to help out.

File: tests/testthat/test-classify-diabetes.R (hunk 5 above)
Comment (Aastedet) [reply tests/testthat/test-classify-diabetes.R:32]: I like loops for their expressiveness/readability, but that's subjective (and not very R'esque). I'd rather be without either though if we can really do it that simple. The reason I went with the loop was to be able to generate specific error messages depending on what went wrong. I was not sure if the default output from the straight-forward expect_identical() would contain that information.
